blocks/library/shortcode/block.js

```diff
@@ -0,0 +1,102 @@
+/**
+ * External dependencies
+ */
+import { connect } from 'react-redux';
+
+/**
+ * WordPress dependencies
+ */
+import { withInstanceId, Dashicon } from '@wordpress/components';
+import { Component, compose } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import ShortcodePreview from './preview';
+import BlockControls from '../../block-controls';
+import { getCurrentPostId } from '../../../editor/store/selectors';
+import PlainText from '../../plain-text';
+
+export class Shortcode extends Component {
+	constructor() {
+		super();
+		this.state = {
+			preview: false,
+		};
+	}
+
+	render() {
+		const { preview } = this.state;
+		const { instanceId, postId, setAttributes, attributes, focus, setFocus } = this.props;
+		const inputId = `blocks-shortcode-input-${ instanceId }`;
+
+		const controls = focus && (
+			<BlockControls key="controls">
+				<div className="components-toolbar">
+					<button
+						className={ `components-tab-button ${ ! preview ? 'is-active' : '' }` }
+						onClick={ () => this.setState( { preview: false } ) }>
+						<span>{ __( 'Shortcode' ) }</span>
+					</button>
+					<button
+						className={ `components-tab-button ${ preview ? 'is-active' : '' }` }
+						onClick={ () => this.setState( { preview: true } ) }>
+						<span>{ __( 'Preview' ) }</span>
+					</button>
+				</div>
+			</BlockControls>
+		);
+
+		if ( ! preview ) {
+			return [
+				controls,
+				<div className="wp-block-shortcode" key="placeholder">
+					<label htmlFor={ inputId }>
+						<Dashicon icon="editor-code" />
+						{ __( 'Shortcode' ) }
+					</label>
+					<PlainText
+						id={ inputId }
+						value={ attributes.text }
+						placeholder={ __( 'Write shortcode here…' ) }
+						onChange={ ( text ) => setAttributes( { text } ) }
+					/>
+				</div>,
+			];
+		}
+
+		const shortcodeContent = ( !! attributes.text ) ? attributes.text.trim() : '';
+
+		if ( ! shortcodeContent.length ) {
+			return [
+				controls,
+				<div key="empty" className="wp-block-embed is-loading">
+					{ __( 'Enter something to preview' ) }
+				</div>,
+			];
+		}
+
+		return [
+			controls,
+			<ShortcodePreview key="preview"
```

Comment:
This is the only place where `postId` is necessary. Can we wrap `ShortcodePreview` with the `connect` HOC only? This will make sure you don't need to tweak unit tests, as well as it assures you don't need to pass down props to the child component.

Reply:
Great suggestion @gziolo. I will move the connect HOC to `ShortcodePreview` so that we can avoid having to tweak the unit tests.

Reply:
@gziolo , this might actually not work out. The postID is actually needed by `withAPIData` HOC in `ShortcodePreview `, so it is essential that it is passed down as a prop. Unless I'm missing out on a way by which the withAPIData can still access the postId from connect HOC.

Reply:
You should be able to do it using the following code:

```js
export default compose( [
	applyConnect,
	applyWithAPIData,
] )( ShortcodePreview );
```

where `applyWithAPIData` is what you already have in `ShortcodePreview` and `applyConnect` needs to be moved to this file. There is no need to pass down `postId` from the parent component.

Reply:
I see it works, awesome 👍